protocol/consistencychecks.go

```diff
@@ -0,0 +1,265 @@
+// Implements all consistency check operations done by a CONIKS client
+// on data received from a CONIKS directory.
+// These include data binding proof verification,
+// and non-equivocation checks.
+
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	m "github.com/coniks-sys/coniks-go/merkletree"
+)
+
+const (
+	invalidProof int = iota
+	proofOfAbsence
+	proofOfInclusion
+)
+
+// ConsistencyChecks stores the current consistency check
+// state of a CONIKS client. This includes the current epoch
+// and SignedTreeRoot value, as well as directory's policies
+// (e.g., wherether the TemporaryBinding extension is being used or not)
+// The client should create a new ConsistencyChecks instance only once,
+// when she registers her binding with a ConiksDirectory.
+// This ConsistencyChecks instance then will be used to verify
+// the returned responses from the ConiksDirectory.
+type ConsistencyChecks struct {
+	CurrentEpoch uint64
+	SavedSTR     []byte
+
+	// extensions settings
+	isUseTBs bool
+	TBs      []*m.TemporaryBinding
+
+	// signing key
+	// TODO: should we pin the signing key in the client? see #47
+	signKey sign.PublicKey
+}
+
+// NewCC creates an instance of ConsistencyChecks using
+// the pinning directory's STR at epoch 0.
+func NewCC(savedSTR []byte, isUseTBs bool, signKey sign.PublicKey) *ConsistencyChecks {
+	return &ConsistencyChecks{
+		CurrentEpoch: 0,
+		SavedSTR:     savedSTR,
+		isUseTBs:     isUseTBs,
+		signKey:      signKey,
+	}
+}
+
+// Verify verifies the directory's response for a request.
+// It first verifies the directory's returned status code of the request.
+// If the status code is not in ErrorResponses array, it means
+// the directory has successfully handled the request.
+// The verifier will then verify the consistency state of the response,
+// and finally, it verifies the returned proof type
+// according to the request type.
+// This will also update the consistency state if the check is passed.
+func (cc *ConsistencyChecks) Verify(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+
+	var verificationResult error
+	var proofType int
+
+	switch msg.DirectoryResponse.(type) {
+	case *DirectoryProof:
+		proofType, verificationResult = cc.verifyDirectoryProof(
+			msg.DirectoryResponse.(*DirectoryProof), uname, key)
+	case *DirectoryProofs:
+		proofType, verificationResult = cc.verifyDirectoryProofs(
+			msg.DirectoryResponse.(*DirectoryProofs), uname, key)
+	}
+
+	if verificationResult != nil {
+		return verificationResult.(ErrorCode)
+	}
+
+	switch requestType {
+	case RegistrationType:
+		if msg.Error == ErrorNameExisted &&
```

Comment:
No, `ErrorResponses` just includes ErrorMalformedClientMessage and ErrorDirectory (see error.go)